pkg/controllers/tagging/tagging_controller.go

```diff
@@ -0,0 +1,248 @@
+		AddFunc:    func(obj interface{}) { tc.enqueueNode(obj, tc.tagNodesResources) },
+		UpdateFunc: func(oldObj, newObj interface{}) { tc.enqueueNode(newObj, tc.tagNodesResources) },
+		DeleteFunc: func(obj interface{}) { tc.enqueueNode(obj, tc.untagNodeResources) },
+	})
+
+	return tc, nil
+}
+
+// Run will start the controller to tag resources attached to the cluster
+// and untag resources detached from the cluster.
+func (tc *TaggingController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer tc.workqueue.ShutDown()
+
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, tc.nodesSynced); !ok {
+		klog.Errorf("failed to wait for caches to sync")
+		return
+	}
+
+	klog.Infof("Starting the tagging controller")
+	go wait.Until(tc.work, tc.nodeMonitorPeriod, stopCh)
+
+	<-stopCh
+}
+
+// work is a long-running function that continuously
+// call process() for each message on the workqueue
+func (tc *TaggingController) work() {
+	for tc.Process() {
+	}
+}
+
+// Process reads each message in the queue and performs either
+// tag or untag function on the Node object
+func (tc *TaggingController) Process() bool {
+	obj, shutdown := tc.workqueue.Get()
+	if shutdown {
+		return false
+	}
+
+	klog.Infof("Starting to process %v", obj)
+
+	err := func(obj interface{}) error {
+		defer tc.workqueue.Done(obj)
+
+		workItem, ok := obj.(*workItem)
+		if !ok {
+			tc.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected workItem in workqueue but got %#v", obj))
+			return nil
+		}
+
+		err := workItem.action(workItem.node)
+		if err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			tc.workqueue.AddRateLimited(workItem)
+			return fmt.Errorf("error finishing work item '%v': %s, requeuing", workItem, err.Error())
+		}
+
+		tc.workqueue.Forget(obj)
+		klog.Infof("Finished processing %v", workItem)
+		return nil
+	}(obj)
+
+	if err != nil {
+		klog.Errorf("Error occurred while processing %v", obj)
+		utilruntime.HandleError(err)
+	}
+
+	return true
+}
+
+// tagNodesResources tag node resources from a list of nodes
+// If we want to tag more resources, modify this function appropriately
+func (tc *TaggingController) tagNodesResources(node *v1.Node) error {
+	for _, resource := range tc.resources {
+		switch resource {
+		case opt.Instance:
+			err := tc.tagEc2Instance(node)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// tagEc2Instances applies the provided tags to each EC2 instance in
+// the cluster.
+func (tc *TaggingController) tagEc2Instance(node *v1.Node) error {
+	instanceId, err := awsv1.KubernetesInstanceID(node.Spec.ProviderID).MapToAWSInstanceID()
+
+	if err != nil {
+		klog.Errorf("Error in getting instanceID for node %s, error: %v", node.GetName(), err)
+		return err
+	} else {
+		err := tc.cloud.TagResource(string(instanceId), tc.tags)
```

Comment:
Are retries and throttling type errors handled somewhere?  I see that retriable errors are handled [here](https://github.com/kubernetes/cloud-provider-aws/blob/master/pkg/providers/v1/tags.go#L227-L238) so as long as we are using that function under the covers then we are ok.  Not sure if we are.

Reply:
it is handled [here](https://github.com/kubernetes/cloud-provider-aws/pull/308/files#diff-ac633e66d7cfdfa6934f05558a33311d493f704ab69e5443c4ad046a47ae1266R145) by simply requeuing the items back to the queue with rate limiting.